File: Delete-Operation-for-Two-Strings/DeleteOperationForTwoStrings.java
class Solution {
    int dp[][];
    public int minDistance(String word1, String word2) {
        dp = new int[word1.length()+1][word2.length()+1];
        for(int[] temp : dp){
            Arrays.fill(temp,-1);
        }
        
        int common = lcs(word1, word2, word1.length(), word2.length());
        return (word1.length()-common)+(word2.length()-common);
    }
    
    int lcs(String s1, String s2, int len1, int len2){
        if(len1 == 0 || len2 == 0) return 0;
        if(dp[len1][len2] != -1) return dp[len1][len2];
        else {
            if(s1.charAt(len1-1) == s2.charAt(len2-1)){
//                list.add(s1.charAt(len1-1));
                dp[len1][len2] = 1 + lcs(s1,s2, len1 - 1, len2 - 1);
            }
            else{
                dp[len1][len2]=Math.max(lcs(s1,s2,len1-1,len2),lcs(s1,s2,len1,len2-1));
            }

            return dp[len1][len2];
        }
    }
}

File: Delete-Operation-for-Two-Strings/Delete-Operation-for-Two-Strings.js
var minDistance = function(word1, word2) {
   /* The idea is to calculate longest common subsequene between both the strings
    * so that excluding those characters, remaining all characters need to be removed.
    * Hence the logic breaks down for removing minimum number of characters as follows:
    *  word1.length + word2.length - 2 * LCS length (for word1 + word2).
    */
    const lcs = Array.from(Array(word1.length + 1), () => Array(word2.length + 1).fill(0));
    const firstWordLength = word1.length;
    const secondWordLength = word2.length;

    for (let m = 1; m <= firstWordLength; m++) {
        for (let n = 1; n <= secondWordLength; n++) {
            const isSameCharacter = word1[m - 1] === word2[n - 1];
            const maxSubsequenceTillNow = Math.max(lcs[m][n - 1], lcs[m - 1][n]);
            lcs[m][n] = isSameCharacter ? 1 + lcs[m - 1][n - 1] : maxSubsequenceTillNow;
        }
    }
    
    return firstWordLength + secondWordLength - 2 * lcs[firstWordLength][secondWordLength];
};
